drivers/gpio/gpio_fxl6408.c

```diff
@@ -0,0 +1,436 @@
+	struct gpio_fxl6408_drv_data *const drv_data =
+		(struct gpio_fxl6408_drv_data *const)dev->data;
+	const struct device *i2c_master = drv_data->i2c_master;
+	int ret = 0;
+
+	ret = i2c_reg_write_byte(i2c_master, config->i2c_slave_addr, reg,
+					 value);
+	if (ret == 0) {
+		*cache = value;
+	} else {
+		LOG_ERR("error writing to register 0x%X (%d)", reg,
+			ret);
+	}
+	LOG_DBG("Write: REG[0x%X] = 0x%X", reg, *cache);
+
+	return ret;
+}
+
+static inline int update_input_regs(const struct device *dev, uint8_t *buf)
+{
+	struct gpio_fxl6408_drv_data *const drv_data =
+		(struct gpio_fxl6408_drv_data *const)dev->data;
+	int ret = read_port_regs(dev, REG_INPUT_VALUE,
+				 &drv_data->reg_cache.input);
+	*buf = drv_data->reg_cache.input;
+	return ret;
+}
+
+static inline int update_output_regs(const struct device *dev, uint8_t value)
+{
+	struct gpio_fxl6408_drv_data *const drv_data =
+		(struct gpio_fxl6408_drv_data *const)dev->data;
+
+	return write_port_regs(dev, REG_OUTPUT, &drv_data->reg_cache.output,
+			value);
+}
+
+static inline int update_direction_regs(const struct device *dev, uint8_t value)
+{
+	struct gpio_fxl6408_drv_data *const drv_data =
+		(struct gpio_fxl6408_drv_data *const)dev->data;
+
+	/* If there is a pin configured as output, disable HIGH-Z mode for that pin.
+	 *  HIGH-Z is currently not supported in the driver
+	 */
+	if (value != 0) {
+		uint8_t *cache = 0;
+		int ret = write_port_regs(dev, REG_OUTPUT_HIGH_Z, cache, ~value);
+
+		if (ret != 0) {
+			LOG_ERR("error setting HIGH-Z");
+		}
+	}
+
+	return write_port_regs(dev, REG_DIRECTION, &drv_data->reg_cache.dir,
+				   value);
+}
+
+static inline int update_pul_sel_regs(const struct device *dev, uint8_t value)
+{
+	struct gpio_fxl6408_drv_data *const drv_data =
+		(struct gpio_fxl6408_drv_data *const)dev->data;
+
+	return write_port_regs(dev, REG_PUD_SEL, &drv_data->reg_cache.pud_sel,
+				   value);
+}
+
+static inline int update_pul_en_regs(const struct device *dev, uint8_t value)
+{
+	struct gpio_fxl6408_drv_data *const drv_data =
+		(struct gpio_fxl6408_drv_data *const)dev->data;
+
+	return write_port_regs(dev, REG_PUD_EN, &drv_data->reg_cache.pud_en,
+				   value);
+}
+
+static int setup_pin_dir(const struct device *dev, uint32_t pin, int flags)
+{
+	struct gpio_fxl6408_drv_data *const drv_data =
+		(struct gpio_fxl6408_drv_data *const)dev->data;
+	uint8_t reg_dir = drv_data->reg_cache.dir;
+	uint8_t reg_out = drv_data->reg_cache.output;
+	int ret = 0;
+
+	/* Update the driver data to the actual situation of the FXL6408*/
+	if ((flags & GPIO_OUTPUT)) {
+		if ((flags & GPIO_OUTPUT_INIT_HIGH)) {
+			reg_out |= BIT(pin);
+		} else if ((flags & GPIO_OUTPUT_INIT_LOW)) {
+			reg_out &= ~BIT(pin);
+		}
+		reg_dir |= BIT(pin);
+	} else {
+		reg_dir &= ~BIT(pin);
+	}
+
+	ret = update_output_regs(dev, reg_out);
+	if (ret != 0) {
+		return ret;
+	}
+
+	ret = update_direction_regs(dev, reg_dir);
+	return ret;
+}
+
+int gpio_fxl6408_init(const struct device *device)
```

Comment:
We traditionally keep _init function toward the bottom of the file. Just above the `struct` defining API functions. Let's move it there.